src/test/resources/imp2.kore

```diff
@@ -0,0 +1,2340 @@
+              merge{S1,S2}(M2:Map{S1,S2},M1:Map{S1,S2}))))
+   []
+
+   // merge is associative
+
+   axiom{R,S1,S2}
+     \forall{R}(
+        M1:Map{S1,S2},
+        \forall{R}(
+           M2:Map{S1,S2},
+           \forall{R}(
+              M3:Map{S1,S2},
+              \equals{Map{S1,S2},R}(
+                 merge{S1,S2}(merge{S1,S2}(M1:Map{S1,S2},M2:Map{S1,S2}),M3:Map{S1,S2}),
+                 merge{S1,S2}(M1:Map{S1,S2},merge{S1,S2}(M2:Map{S1,S2},M3:Map{S1,S2}))))))
+   []
+
+   // merge is disjoint
+
+   axiom{R,S1,S2}
+     \forall{R}(
+        K:S1,
+        \forall{R}(
+           V1:S2,
+           \forall{R}(
+              V2:S2,
+              \equals{Map{S1,S2},R}(
+                 merge{S1,S2}(bind{S1,S2}(K:S1,V1:S2),bind{S1,S2}(K:S1,V2:S2)),
+                 \bottom{Map{S1,S2}}))))
+   []
+
+   /* lookup, update, and insert will be hooked to external libraries. */
+   symbol lookup{S1,S2}(Map{S1,S2}, S1) : S2 []
+
+   symbol update{S1,S2}(Map{S1,S2}, S1, S2) : Map{S1,S2} []
+
+   symbol insert{S1,S2}(Map{S1,S2}, S1, S2) : Map{S1,S2} []
+
+endmodule []
+
+module ID
+
+  import INT []
+
+  sort Id{} []
+
+  symbol id{}(Int{}) : Id{} []
+
+  /* id is injective. */
+
+  axiom{}
+    \implies{Id{}}(
+      \and{Id{}}(id{}(I1:Int{}), id{}(I2:Int{})),
+      id{}(\and{Int{}}(I1:Int{}, I2:Int{})))
+  []
+
+endmodule []
+
+module CONTEXT
+
+  sort Ctxt{S1,S2} []
+
+  /* Context application are often written as _[_]. */
+  symbol app{S1,S2}(Ctxt{S1,S2}, S1) : S2 []
+
+  symbol gamma{S1,S2}(S1,S2) : Ctxt{S1,S2} []
+
+  /* app and gamma are functional symbols. */
+
+  axiom{S1,S2,R}
+    \forall{R}(
+      C:Ctxt{S1,S2},
+      \forall{R}(
+        T:S1,
+	\exists{R}(
+	  Z:S2,
+	  \equals{S2,R}(
+	    app{S1,S2}(C:Ctxt{S1,S2}, T:S1),
+	    Z:S2))))
+  []
+
+  axiom{S1,S2,R}
+    \forall{R}(
+      HOLE:S1,
+      \forall{R}(
+        BODY:S2,
+	\exists{R}(
+	  C:Ctxt{S1,S2},
+	  \equals{Ctxt{S1,S2},R}(
+	    gamma{S1,S2}(HOLE:S1, BODY:S2),
+	    C:Ctxt{S1,S2}))))
+  []
+
+  /* Identity context. */
+  symbol idctxt{S}() : Ctxt{S,S} []
+
+  /** Axioms about contexts have monad-style.
+    * That is, there are two axioms about contexts as follows.
+    * (Unit) idctxt[T] = T
+    * (Multiplication) C1[C2[T]] = C[T] where C = gamma HOLE . C1[C2[HOLE]]
+    */
+
+  axiom{S,R}
+    \equals{S,R}(
+      app{S,S}(idctxt{S}(), T:S), T:S)
+  []
+
+  axiom{S,S1,S2,R}
+    \equals{S,R}(
+      app{S1,S}(
+        C1:Ctxt{S1,S},
+	app{S2,S1}(
+	  C2:Ctxt{S2,S1},
+	  T:S2)),
+      app{S2,S}(
+        gamma{S2,S}(
+	  HOLE:S2,
+          app{S1,S}(
+            C1:Ctxt{S1,S},
+	    app{S2,S1}(
+	      C2:Ctxt{S2,S1},
+	      HOLE:S2))),
+	T:S2))
+  []
+
+endmodule []
+
+/** The module of "~>" and "." in K.
+  */
+module ARROW
+
+  import CONTEXT []
+
+  sort K{} []
+
+  /* The injection function from any sort S to K. */
+  symbol asK{S}(S) : K{} []
+
+  /* Two injectivity axioms:
+   *   not (asK{S1}(X:S1) /\ asK{S2}(Y:S2))

```

Comment:
The same question as I asked in Xiaohong's pull request:

Shouldn't this require that S1 and S2 are different? I'm not sure how to do that, though.